node/impl/full/trace.go

```diff
@@ -0,0 +1,515 @@
+		}
+
+		traceBlocks := make([]*TraceBlock, 0, len(trace))
+		for _, trace := range traces {
+			traceBlocks = append(traceBlocks, &TraceBlock{
+				Trace:               trace,
+				BlockHash:           blkHash,
+				BlockNumber:         int64(ts.Height()),
+				TransactionHash:     *txHash,
+				TransactionPosition: idx,
+			})
+		}
+
+		allTraces = append(allTraces, traceBlocks...)
+	}
+
+	return allTraces, nil
+}
+
+func (e *EthTrace) TraceReplayBlockTransactions(ctx context.Context, blkNum string, traceTypes []string) (interface{}, error) {
+	if len(traceTypes) != 1 || traceTypes[0] != "trace" {
+		return nil, fmt.Errorf("only 'trace' is supported")
+	}
+
+	ts, err := e.getTipsetByBlockNr(ctx, blkNum, false)
+	if err != nil {
+		return nil, err
+	}
+
+	_, trace, err := e.StateManager.ExecutionTrace(ctx, ts)
+	if err != nil {
+		return nil, xerrors.Errorf("failed when calling ExecutionTrace: %w", err)
+	}
+
+	allTraces := make([]*TraceReplayBlockTransaction, 0, len(trace))
+	for _, ir := range trace {
+		// ignore messages from f00
+		if ir.Msg.From.String() == "f00" {
+			continue
+		}
+
+		txHash, err := e.EthGetTransactionHashByCid(ctx, ir.MsgCid)
+		if err != nil {
+			return nil, err
+		}
+		if txHash == nil {
+			log.Warnf("cannot find transaction hash for cid %s", ir.MsgCid)
+			continue
+		}
+
+		t := TraceReplayBlockTransaction{
+			Output:          hex.EncodeToString(ir.MsgRct.Return),
+			TransactionHash: *txHash,
+			StateDiff:       nil,
+			VmTrace:         nil,
+		}
+
+		err = buildTraces(&t.Trace, nil, []int{}, ir.ExecutionTrace, int64(ts.Height()))
+		if err != nil {
+			return nil, xerrors.Errorf("failed when building traces: %w", err)
+		}
+
+		allTraces = append(allTraces, &t)
+	}
+
+	return allTraces, nil
+}
+
+func writePadded[T any](w io.Writer, data T, size int) error {
+	tmp := &bytes.Buffer{}
+
+	// first write data to tmp buffer to get the size
+	err := binary.Write(tmp, binary.BigEndian, data)
+	if err != nil {
+		return fmt.Errorf("writePadded: failed writing tmp data to buffer: %w", err)
+	}
+
+	if tmp.Len() > size {
+		return fmt.Errorf("writePadded: data is larger than size")
+	}
+
+	// write tailing zeros to pad up to size
+	cnt := size - tmp.Len()
+	for i := 0; i < cnt; i++ {
+		err = binary.Write(w, binary.BigEndian, uint8(0))
+		if err != nil {
+			return fmt.Errorf("writePadded: failed writing tailing zeros to buffer: %w", err)
+		}
+	}
+
+	// finally write the actual value
+	err = binary.Write(w, binary.BigEndian, tmp.Bytes())
+	if err != nil {
+		return fmt.Errorf("writePadded: failed writing data to buffer: %w", err)
+	}
+
+	return nil
+}
+
+// buildTraces recursively builds the traces for a given ExecutionTrace by walking the subcalls
+func buildTraces(traces *[]*Trace, parent *Trace, addr []int, et types.ExecutionTrace, height int64) error {
+	trace := &Trace{
+		Action: Action{
+			From:    et.Msg.From.String(),
+			To:      et.Msg.To.String(),
```

Comment:
We should be converting these to Eth addresses. I.e. https://github.com/filecoin-project/lotus/blob/cbbb76b2f095782e5c4d717b5c2603ccd41d21ff/chain/types/ethtypes/eth_types.go#L339

Reply:
Hm. Also, this is going to be tricky.

1. For sub-calls, the "from" address will always be an ID address. But this is going to be _very_ confusing as users will expect f410f addresses (I.e., the "true" eth address of a contract).
2. These addresses may not be convertible into Eth-style addresses. I.e., the function I linked above may fail. In that case, we'd need to lookup the associated ID address of the actor, then convert _that_ into an eth-style address (which will always work).

There are several subtitles here. I'll think on this and get back to you tomorrow.

Reply:
https://github.com/filecoin-project/lotus/blob/cbbb76b2f095782e5c4d717b5c2603ccd41d21ff/node/impl/full/eth.go#L2047 will be helpful.

Reply:
From what I can tell, yes.

Unless the contract in question somehow was created and deleted in the same epoch. Which isn't currently possible (as far as I know) but.... eh, it's probably fine to fail in that case and we can handle it later.

Reply:
> Unless the contract in question somehow was created and deleted in the same epoch. Which isn't currently possible (as far as I know) but.... eh, it's probably fine to fail in that case and we can handle it later.

Ah... ok, so, this can happen:

1. A message creates a contract.
2. Stuff happens.
3. The top-level message gets reverted due to some failure.

In this case, the contract won't exist in state so `lookupEthAddress` may fail.

So................

First, I think we'll need to add some form of "actor resolved" event to the trace. E.g., replace `Invoke(Cid)` with `ActorResolved(ActorID, CodeCID)`. That ensures that we _always_ have enough information to fill out the trace.

Then:

1. When parsing the trace, remember the address on which the _caller_ was called and always _try_ use that as the "from" address (using `EthAddressFromFilecoinAddress`).
2. If that doesn't work, convert the ID from the trace.
3. For the top-level sender, we'll have to use `lookupEthAddress`.

Reply:
now using `lookupEthAddress`, but left a TODO for supporting the error case, for now, if an error occurs when calling `lookupEthAddress` then we fail that trace only but otherwise continue